File: easy/100_Same_Tree.py
# Given the roots of two binary trees p and q, write a function to check if they are the same or not.
#
# Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        val_list1 = list()
        if not p and not q:
            return True
        if not p or not q:
            return False

        def rec(node: TreeNode):
            if node.left:
                val_list1.append('left')
                rec(node.left)
            val_list1.append(node.val)
            if node.right:
                val_list1.append('right')
                rec(node.right)
        rec(p)
        val_list2 = val_list1[:]
        val_list1 = list()
        rec(q)
        return True if val_list2 == val_list1 else False

# решение без рекурсии
# class Solution(object):
#     def isSameTree(self, p, q):
#         """
#         :type p: TreeNode
#         :type q: TreeNode
#         :rtype: bool
#         """
#         stack1 = [];
#         stack1.append(p)
#         stack2 = [];
#         stack2.append(q)
#
#         while len(stack1) and len(stack2):
#             node1 = stack1.pop(0)
#             node2 = stack2.pop(0)
#             if (not node1 and node2) or (node1 and not node2):
#                 return False
#             elif not node1 and not node2:
#                 continue
#             elif node1.val != node2.val:
#                 return False
#             stack1.append(node1.left)
#             stack1.append(node1.right)
#             stack2.append(node2.left)
#             stack2.append(node2.right)
#         return True

File: go/easy/100_Same_Tree.go
package main

import "fmt"

//Given the roots of two binary trees p and q, write a function to check if they are the same or not.
//Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.

//Example 1:
//Input: p = [1,2,3], q = [1,2,3]
//Output: true
//
//Example 2:
//Input: p = [1,2], q = [1,null,2]
//Output: false

// Example 3:
// Input: p = [1,2,1], q = [1,1,2]
// Output: false

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isSameTree(p *TreeNode, q *TreeNode) bool {
	if p == nil && q == nil {
		return true
	}
	if p == nil || q == nil {
		return false
	}
	if p.Val != q.Val {
		return false
	}
	return isSameTree(p.Left, q.Left) && isSameTree(p.Right, q.Right)
}

func main() {
	p := &TreeNode{
		Val: 5,
		Left: &TreeNode{
			Val: 2,
			Left: &TreeNode{
				Val: 1,
				Right: &TreeNode{
					Val: 3,
				},
			},
		},
		Right: &TreeNode{
			Val: 8,
			Left: &TreeNode{
				Val: 6,
			},
			Right: &TreeNode{
				Val: 10,
			},
		},
	}
	q := &TreeNode{
		Val: 5,
		Left: &TreeNode{
			Val: 2,
			Left: &TreeNode{
				Val: 1,
				Right: &TreeNode{
					Val: 3,
				},
			},
		},
		Right: &TreeNode{
			Val: 8,
			Left: &TreeNode{
				Val: 6,
			},
			Right: &TreeNode{
				Val: 9,
			},
		},
	}
	fmt.Println(isSameTree(p, q))
}
